File: kakao.app/js/ked/view/status/statusfile.js
var _k_ = {extend: function (c,p) {for (var k in p) { if (Object.hasOwn(p, k)) c[k] = p[k] } function ctor() { this.constructor = c; } ctor.prototype = p.prototype; c.prototype = new ctor(); c.__super__ = p.prototype; return c;}, list: function (l) {return l != null ? typeof l.length === 'number' ? l : [] : []}, in: function (a,l) {return (typeof l === 'string' && typeof a === 'string' && a.length ? '' : []).indexOf.call(l,a) >= 0}, trim: function (s,c=' ') {return _k_.ltrim(_k_.rtrim(s,c),c)}, ltrim: function (s,c=' ') { while (_k_.in(s[0],c)) { s = s.slice(1) } return s}, rtrim: function (s,c=' ') {while (_k_.in(s.slice(-1)[0],c)) { s = s.slice(0, s.length - 1) } return s}}

var statusfile

import kxk from "../../../kxk.js"
let slash = kxk.slash
let post = kxk.post

import color from "../../theme/color.js"
import theme from "../../theme/theme.js"

import syntax from "../../util/syntax.js"

import view from "../base/view.js"


statusfile = (function ()
{
    _k_.extend(statusfile, view)
    function statusfile (screen, name)
    {
        this["onMouse"] = this["onMouse"].bind(this)
        this["show"] = this["show"].bind(this)
        statusfile.__super__.constructor.call(this,screen,name)
        this.setColor('bg',theme.status.bg)
        this.setColor('empty',theme.status.empty)
        this.setColor('hover',theme.hover.bg)
        this.pointerType = 'pointer'
        this.syntax = new syntax
        this.syntax.setExt('noon')
        this.rounded = ''
    }

    statusfile.prototype["draw"] = function ()
    {
        var bg, ch, fg, x

        if (this.hidden())
        {
            return
        }
        statusfile.__super__.draw.call(this)
        bg = (this.hover ? this.color.hover : this.color.bg)
        var list = _k_.list(this.rounded)
        for (x = 0; x < list.length; x++)
        {
            ch = list[x]
            if (_k_.in(ch,''))
            {
                this.cells.set(x,0,ch,bg,this.color.empty)
            }
            else
            {
                fg = this.syntax.getColor(x,0)
                this.cells.set(x,0,ch,fg,bg)
                if (this.hover)
                {
                    this.cells.adjustContrastForHighlight(x,0,bg)
                }
            }
        }
    }

    statusfile.prototype["adjustText"] = function ()
    {
        var _62_14_

        this.file = ((_62_14_=this.file) != null ? _62_14_ : '')
        this.pars = slash.parse(this.file)
        this.syntax.clear()
        this.syntax.setLines(['/' + this.pars.file])
        return this.rounded = '' + this.pars.file + ''
    }

    statusfile.prototype["set"] = function (file)
    {
        if (this.file === _k_.trim(file))
        {
            return
        }
        this.file = _k_.trim(file)
        return this.adjustText()
    }

    statusfile.prototype["show"] = function (file)
    {
        this.set(file)
        return this.cells.rows = 1
    }

    statusfile.prototype["onMouse"] = function (event)
    {
        statusfile.__super__.onMouse.call(this,event)
    
        switch (event.type)
        {
            case 'press':
                if (this.hover)
                {
                    this.emit('action','click',this.file)
                    return {redraw:true}
                }
                break
        }

        return this.hover
    }

    return statusfile
})()

export default statusfile;

File: kakao.app/love/lua/view/editor/statusfile.lua
--[[
     ███████  █████████   ███████   █████████  ███   ███   ███████  ████████  ███  ███      ████████  
    ███          ███     ███   ███     ███     ███   ███  ███       ███       ███  ███      ███       
    ███████      ███     █████████     ███     ███   ███  ███████   ██████    ███  ███      ███████   
         ███     ███     ███   ███     ███     ███   ███       ███  ███       ███  ███      ███       
    ███████      ███     ███   ███     ███      ███████   ███████   ███       ███  ███████  ████████  

    used by status to display the currently edited file in the fileeditor
--]]

view = require "view.base.view"
syntax = require "util.syntax"


local statusfile = class("statusfile", view)
    


function statusfile:init(name) 
        view.init(self, name)
        
        self:setColor('bg', theme.status.bg)
        self:setColor('empty', theme.status.empty)
        self:setColor('hover', theme.hover.bg)
        
        self.pointerType = 'pointer'
        self.syntax = syntax()
        self.syntax:setExt('noon')
        
        self.rounded = array('', '')
        return self
    end

-- 0000000    00000000    0000000   000   000  
-- 000   000  000   000  000   000  000 0 000  
-- 000   000  0000000    000000000  000000000  
-- 000   000  000   000  000   000  000   000  
-- 0000000    000   000  000   000  00     00  


function statusfile:draw() 
        if self:hidden() then return end
        
        -- view.draw @
        
        local bg = (function () 
    if self.hover then 
    return self.color.hover else 
    return self.color.bg
             end
end)()
        
        for x, ch in ipairs(self.rounded) do 
            if ((ch == '') or (ch == '')) then 
                self.cells:set(x, 1, ch, bg, self.color.empty)
            else 
                local fg = self.syntax:getColor(x, 1)
                self.cells:set(x, 1, ch, fg, bg)
                if self.hover then 
                    self.cells:adjustContrastForHighlight(x, 1, bg)
                end
            end
        end
        
        return self:render()
    end

--  0000000   0000000          000  000   000   0000000  000000000  
-- 000   000  000   000        000  000   000  000          000     
-- 000000000  000   000        000  000   000  0000000      000     
-- 000   000  000   000  000   000  000   000       000     000     
-- 000   000  0000000     0000000    0000000   0000000      000     


function statusfile:adjustText() 
        self.file = self.file or ''
        self.pars = slash.parse(self.file)
        self.syntax:clear()
        self.syntax:setLines(array('/' .. self.pars.file))
        self.rounded = ((array('') + kseg(self.pars.file)) + array(''))
        return self.rounded
    end


function statusfile:set(file) 
        if (self.file == kstr.trim(file)) then return end
        
        self.file = kstr.trim(file)
        return self:adjustText()
    end


function statusfile:show(file) 
        self:set(file)
        self.cells.rows = 1
        return self.cells.rows
    end

-- 00     00   0000000   000   000   0000000  00000000  
-- 000   000  000   000  000   000  000       000       
-- 000000000  000   000  000   000  0000000   0000000   
-- 000 0 000  000   000  000   000       000  000       
-- 000   000   0000000    0000000   0000000   00000000  


function statusfile:onMouse(event) 
        view.onMouse(self, event)
        
        if (event.type == 'press') then 
                if self.hover then 
                    self:emit('action', 'click', self.file)
                    return true
                end
        end
        
        return self.hover
    end

return statusfile